File: backend/services/ai_assessment_service.py
"""
services/ai_assessment_service.py

Orchestration layer for AI Agent routes. For this proof-of-concept slice,
it wires the Difficulty Engine and QuestionGenerationAgent. As the
remaining agents in ARCHITECTURE.md are built (Assessment Planner,
Quality Validation, Assessment Builder, ...), this file grows into the
place where they're chained together — routes/ai_assessment_routes.py
should never need to change shape when that happens, only this file.
"""

from agents.question_generation_agent import (
    QuestionGenerationAgent,
    QuestionGenerationError,
)
from firebase.firebase_config import get_firestore_client
from services.difficulty_engine import compute_difficulty, DifficultyDecision
from services.assessment_planner import build_diagnostic_plan
from services.evaluation_service import evaluate_diagnostic_assessment


class AIAssessmentError(Exception):
    pass


def resolve_difficulty(signals: dict) -> DifficultyDecision:
    """
    Runs the Difficulty Engine on raw student performance signals.
    Kept as its own function (not inlined into generate_ai_questions) so
    a future route can call it standalone — e.g. to show "why did I get
    Hard questions?" in the UI before generation even happens.
    """
    return compute_difficulty(
        previous_score=float(signals.get("previous_score", 50)),
        time_taken_seconds=float(signals.get("time_taken_seconds", 0)),
        expected_time_seconds=float(signals.get("expected_time_seconds", 0)),
        confidence=float(signals.get("confidence", 50)),
        mistake_rate=float(signals.get("mistake_rate", 0)),
    )


def generate_ai_questions(
    skill: str,
    topics: list[str],
    count: int,
    difficulty: str | None = None,
    signals: dict | None = None,
    learning_objective: str = "",
) -> dict:
    """
    Runs the Question Generation Agent and returns its output alongside
    the difficulty decision that produced it.

    Exactly one of `difficulty` or `signals` should be provided:
      - `difficulty`: explicit override ("Easy"/"Medium"/"Hard") — used by
        admins/testing, or once the Assessment Planner Agent (§9 Phase 2)
        passes a difficulty it already decided.
      - `signals`: raw student performance data, run through the
        Difficulty Engine to decide difficulty automatically. This is the
        normal path for real assessments.

    Not persisted anywhere — per the architecture rule, AI-generated
    questions are ephemeral until the (future) Assessment Builder Agent
    assembles them into an assessment the student actually takes.
    """
    reasoning = None
    if difficulty:
        pass  # explicit override, no engine involved
    elif signals:
        decision = resolve_difficulty(signals)
        difficulty = decision.difficulty
        reasoning = decision.reasoning
    else:
        raise AIAssessmentError(
            "Provide either 'difficulty' (explicit) or 'signals' "
            "(previous_score, time_taken_seconds, expected_time_seconds, "
            "confidence, mistake_rate) so the Difficulty Engine can decide."
        )

    agent = QuestionGenerationAgent()
    try:
        questions = agent.run(
            topics=topics,
            difficulty=difficulty,
            count=count,
            skill=skill,
            learning_objective=learning_objective,
        )
    except QuestionGenerationError as exc:
        raise AIAssessmentError(str(exc)) from exc

    return {
        "difficulty": difficulty,
        "difficulty_reasoning": reasoning,  # None when explicitly overridden
        "questions": questions,
    }


def generate_diagnostic_assessment(
    skills: list[str], role: str = "", learning_objective: str = ""
) -> dict:
    """
    The real diagnostic assessment: one QuestionGenerationAgent.run_chunked()
    call PER selected skill (via the Assessment Planner's fixed 5 Easy +
    5 Medium + 5 Hard = 15-question plan), aggregated into one question set.

    run_chunked() itself makes 3 smaller Gemini calls per skill (one per
    difficulty, 5 questions each) instead of one 15-question call — see
    agents/question_generation_agent.py and services/assessment_planner.py
    for why: a single big call is more failure-prone, and one skill's
    total failure used to abort the entire diagnostic (5 skills selected
    = 5 chances to crash everything). Chunking shrinks the failure
    surface back down to "one difficulty band of one skill," which
    should now be about as reliable as the old 6-question-total call was.

    Deliberately sequential, not parallel — keeps retry/backoff behavior
    (agents/question_generation_agent.py) simple and predictable, and
    Groq's per-call latency is low enough that even 5-6 skills completes
    well within the frontend's timeout. If this becomes a bottleneck with
    many more skills, parallelizing these calls is a contained change
    right here — nothing else in the stack needs to know.

    Raises AIAssessmentError with a partial-failure message identifying
    which specific skill (and which difficulty chunk within it) failed,
    rather than a generic "something broke".
    """
    plan = build_diagnostic_plan(skills)
    agent = QuestionGenerationAgent()
    all_questions: list[dict] = []

    for skill_plan in plan:
        try:
            questions = agent.run_chunked(
                topics=[skill_plan.skill],
                skill=skill_plan.skill,
                difficulty_counts=skill_plan.difficulty_counts,
                open_ended_counts=skill_plan.open_ended_counts,
                open_ended_type=skill_plan.open_ended_type,
                learning_objective=learning_objective or (f"for the {role} role" if role else ""),
            )
        except QuestionGenerationError as exc:
            raise AIAssessmentError(
                f"Diagnostic assessment generation failed on skill "
                f"'{skill_plan.skill}': {exc}"
            ) from exc

        # CRITICAL: each run_chunked() call numbers its own questions
        # "AI-1".."AI-15" independently — across multiple skills these
        # collide (every skill would have an "AI-1"). Since evaluation
        # matches answers by TempID, colliding IDs silently corrupt
        # scoring (a later skill's answer overwrites an earlier skill's
        # under the same key). Re-namespace by skill right here, once,
        # so every ID in the aggregated set is globally unique.
        for q in questions:
            q["TempID"] = f"{skill_plan.skill}::{q['TempID']}"

        all_questions.extend(questions)

    return {
        "skills": skills,
        "totalQuestions": len(all_questions),
        "questions": all_questions,
    }


def evaluate_assessment(questions: list[dict], answers: dict[str, str]) -> dict:
    """
    Thin wrapper around the Evaluation Agent (services/evaluation_service.py)
    so routes only ever import from ai_assessment_service.py, same as
    every other AI feature — keeps one consistent import surface instead
    of routes reaching into individual agent/service modules directly.
    """
    result = evaluate_diagnostic_assessment(questions, answers)
    return result.to_dict()


def evaluate_and_save_assessment(
    uid: str, role: str, skills: list[str],
    questions: list[dict], answers: dict[str, str],
) -> dict:
    """
    Evaluates AND persists the full result (questions, answers,
    evaluation) so a page refresh loads this saved attempt instead of
    silently generating a brand-new assessment. See
    services/assessment_repository.py for why the FULL result is saved,
    not just the evaluation summary.
    """
    from services.assessment_repository import save_assessment_result

    evaluation = evaluate_assessment(questions, answers)
    db = get_firestore_client()
    save_assessment_result(db, uid, role, skills, questions, answers, evaluation)
    return evaluation


def load_saved_assessment_result(uid: str) -> dict | None:
    """
    Returns the user's last completed assessment (questions, answers,
    evaluation) or None if they haven't completed one yet. None should
    be treated as "show the normal take-the-assessment flow", not an
    error — this is the "don't regenerate on every refresh" fix.
    """
    from services.assessment_repository import get_assessment_result

    db = get_firestore_client()
    return get_assessment_result(db, uid)


def quit_role(uid: str) -> None:
    """
    "Quit Role" (Learning Hub): wipes this student's saved assessment
    AND saved roadmap so Role Selection unlocks again — deliberately
    both, since a leftover roadmap with no matching assessment (or vice
    versa) would leave the app in a half-quit, inconsistent state.
    """
    from services.assessment_repository import delete_assessment_result
    from services.roadmap_repository import delete_roadmap

    db = get_firestore_client()
    delete_assessment_result(db, uid)
    delete_roadmap(db, uid)

File: frontend/src/services/aiAssessmentService.js
import apiClient from "../api/axiosClient";
import { ENDPOINTS } from "../api/endpoints";

/**
 * aiAssessmentService — calls the (real, deployed) Question Generation
 * Agent at POST /api/ai/generate-questions.
 *
 * Uses a longer timeout than the default apiClient (10s) because:
 *  - Render free-tier services cold-start (~15-50s) after idling
 *  - The Gemini call itself + one built-in retry can take several seconds
 * 100s covers cold start + a retried Gemini call with its 2s backoff
 * delay, without the request timing out on a slow first hit — the same
 * class of problem Postman's Cloud Agent hit with its 30s cap.
 */
const GENERATION_TIMEOUT_MS = 100000;

/**
 * @param {object} params
 * @param {string} params.skill - e.g. the selected career role title
 * @param {string[]} params.topics - selected skills, treated as topics
 * @param {number} params.count
 * @param {string} [params.difficulty] - explicit override, skips the
 *   Difficulty Engine ("Easy"|"Medium"|"Hard")
 * @param {object} [params.signals] - raw performance signals for the
 *   Difficulty Engine to decide difficulty from instead (see
 *   backend/services/difficulty_engine.py). Provide this OR difficulty,
 *   not both — the backend requires exactly one.
 */
export async function generateAssessmentQuestions({
  skill,
  topics,
  count = 5,
  difficulty,
  signals,
  learning_objective = "",
}) {
  const payload = { skill, topics, count, learning_objective };
  if (difficulty) payload.difficulty = difficulty;
  else if (signals) payload.signals = signals;

  const { data } = await apiClient.post(
    ENDPOINTS.ASSESSMENT.GENERATE_QUESTIONS,
    payload,
    { timeout: GENERATION_TIMEOUT_MS }
  );

  // Backend response_helper shape: { success, message, data: {...} }
  if (!data.success) {
    throw new Error(data.error || data.message || "Question generation failed.");
  }
  // data.data shape (routes/ai_assessment_routes.py):
  //   { difficulty: "Medium", difficulty_reasoning: "..."|null, questions: [...] }
  return data.data;
}

/**
 * generateDiagnosticAssessment — the real diagnostic flow. One call
 * generates a fixed 5 Easy + 5 Medium + 5 Hard set PER selected skill
 * (backend/services/assessment_planner.py), all aggregated into one
 * question set with globally-unique TempIDs (each prefixed by skill).
 *
 * @param {string[]} skills - the student's selected skills (each gets
 *   its own 6-question Easy/Medium/Hard block)
 * @param {string} [role] - selected career role, used as generation
 *   context only (not a topic)
 * @returns {Promise<{skills: string[], totalQuestions: number, questions: object[]}>}
 */
export async function generateDiagnosticAssessment({ skills, role = "", learning_objective = "" }) {
  const { data } = await apiClient.post(
    ENDPOINTS.ASSESSMENT.GENERATE_DIAGNOSTIC_ASSESSMENT,
    { skills, role, learning_objective },
    { timeout: GENERATION_TIMEOUT_MS }
  );
  if (!data.success) {
    throw new Error(data.error || data.message || "Diagnostic assessment generation failed.");
  }
  return data.data;
}

/**
 * evaluateDiagnosticAssessment — Evaluation Agent. Send back the exact
 * `questions` array generateDiagnosticAssessment returned, plus the
 * student's answers keyed by TempID (skipped questions simply omitted).
 * Returns the skill-wise Strong/Intermediate/Weak/Not Attempted table.
 *
 * Passing `uid`/`role`/`skills` also saves the FULL result (questions,
 * answers, evaluation) to Firestore — this is what makes a page refresh
 * show the same completed result instead of silently regenerating a
 * brand-new assessment. Omit them to evaluate without persisting.
 *
 * @param {object[]} questions
 * @param {Record<string, string>} answers - {TempID: "OptionA" | <typed answer text>, ...}
 * @param {string} [uid]
 * @param {string} [role]
 * @param {string[]} [skills]
 * @returns {Promise<{skills: object[], overall: {correct, total, scorePercent}, questionResults: Record<string, boolean>}>}
 */
export async function evaluateDiagnosticAssessment(questions, answers, uid = null, role = "", skills = []) {
  // NOT "fast, local, no Gemini call involved" — evaluation grades every
  // FillBlank/CodeCompletion answer via answer_equivalence_service.py,
  // which calls Gemini for anything that isn't an exact normalized-string
  // match. Worse, submission happens minutes after generation, with no
  // requests hitting the backend in between while the student is
  // answering — long enough for Render's free tier to spin the backend
  // down, so this request often has to eat a cold start (15-50s) AND
  // the grading calls. EVALUATION_TIMEOUT_MS gives it room for both;
  // the previous 10s default timeout is exactly why submission was
  // intermittently failing after a long assessment.
  const EVALUATION_TIMEOUT_MS = 90000;

  const { data } = await apiClient.post(
    ENDPOINTS.ASSESSMENT.EVALUATE_DIAGNOSTIC_ASSESSMENT,
    { questions, answers, uid, role, skills },
    { timeout: EVALUATION_TIMEOUT_MS }
  );
  if (!data.success) {
    throw new Error(data.error || data.message || "Evaluation failed.");
  }
  return data.data;
}

/**
 * loadSavedAssessmentResult — checks whether this user already has a
 * completed assessment saved. Call this BEFORE generating a new one —
 * if it returns non-null, show that saved result instead of calling
 * generateDiagnosticAssessment at all.
 *
 * @param {string} uid
 * @returns {Promise<{role, skills, questions, answers, evaluation}|null>}
 */
export async function loadSavedAssessmentResult(uid) {
  const { data } = await apiClient.get(`/assessment-result/${uid}`);
  if (!data.success) {
    throw new Error(data.error || data.message || "Failed to load your assessment result.");
  }
  return data.data; // null if none saved yet
}

/**
 * generateRoadmap — Roadmap Agent. Send the exact object
 * evaluateDiagnosticAssessment returned. Rule-based on the backend, not
 * an AI call, so this is fast and never fails due to Gemini/Groq being
 * down — default timeout is fine here too.
 *
 * Passing `uid` (and `role`) saves the roadmap to Firestore so it
 * survives a page reload instead of disappearing — omit it (e.g. for
 * quick local testing) and the roadmap is still generated, just not
 * persisted anywhere.
 *
 * Passing `roleId` (e.g. "frontend" — see constants/roles.js) makes the
 * roadmap role-driven: every skill in that role's syllabus is included,
 * not just the ones assessed — unclaimed skills come back as their own
 * "not_assessed" entries instead of disappearing. Omit it and the
 * roadmap falls back to assessed-skills-only, same as before. Either
 * way, the response now also includes `compressedSyllabus` — the
 * topic-level Verified/Current/Locked tree per skill, ready to hand
 * straight to <RoadmapDisplay compressedSyllabus={...} /> — no second
 * call needed.
 *
 * @param {object} evaluation - {skills: [...], overall: {...}}
 * @param {string} [uid] - Firebase uid, enables persistence
 * @param {string} [role] - selected career role TITLE, stored alongside the roadmap
 * @param {string} [roleId] - selected career role ID, drives the full curriculum
 * @returns {Promise<{entries: object[], totalWeeks: number, includesProjectWeek: boolean, notAssessedCount: number, compressedSyllabus: object|null, currentWeek?: number, completionPercent?: number}>}
 */
export async function generateRoadmap(evaluation, uid = null, role = "", roleId = "") {
  const { data } = await apiClient.post(ENDPOINTS.ASSESSMENT.GENERATE_ROADMAP, {
    evaluation,
    uid,
    role,
    roleId,
  });
  if (!data.success) {
    throw new Error(data.error || data.message || "Roadmap generation failed.");
  }
  return data.data;
}

/**
 * loadSavedRoadmap — fetches a previously generated & saved roadmap for
 * this user, if one exists. Returns null if the user has never taken
 * the diagnostic assessment yet (not an error — a normal, expected state).
 *
 * @param {string} uid
 * @returns {Promise<object|null>}
 */
export async function loadSavedRoadmap(uid) {
  const { data } = await apiClient.get(`/roadmap/${uid}`);
  if (!data.success) {
    throw new Error(data.error || data.message || "Failed to load roadmap.");
  }
  return data.data; // null if none saved yet
}

/**
 * recomputeRoadmapMastery — POSTs to /roadmap/<uid>/recompute, which
 * re-judges every skill on the saved roadmap under the CURRENT mastery
 * rules (services.roadmap_service.recompute_all_mastery). Needed
 * because the automatic recompute only fires as a side-effect of a NEW
 * topic quiz submission — progress completed before a mastery-logic
 * change ships otherwise stays stuck showing the old verdict forever.
 * See services/userProgressCache.js, which calls this once per Profile
 * dashboard load (bounded by the same cache TTL) and invalidates the
 * roadmap cache if anything actually changed.
 *
 * @param {string} uid
 * @returns {Promise<object|null>} the recomputed roadmap, or null if
 *   the user has no saved roadmap yet.
 */
export async function recomputeRoadmapMastery(uid) {
  const { data } = await apiClient.post(`/roadmap/${uid}/recompute`);
  if (!data.success) {
    throw new Error(data.error || data.message || "Failed to recompute roadmap.");
  }
  return data.data;
}

/**
 * quitRole — "Quit Role" (Learning Hub). Deletes the saved assessment
 * AND saved roadmap for this uid so Role Selection unlocks again. This
 * is the ONLY way back to Role Selection once a role has been chosen —
 * there's no other "change course" shortcut left in the app.
 *
 * @param {string} uid
 */
export async function quitRole(uid) {
  const { data } = await apiClient.delete(ENDPOINTS.ASSESSMENT.QUIT_ROLE(uid));
  if (!data.success) {
    throw new Error(data.error || data.message || "Failed to quit role.");
  }
}
